File: src/main/java/net/onelitefeather/stardust/listener/PlayerVanishListener.java
package net.onelitefeather.stardust.listener;

import com.destroystokyo.paper.event.entity.SkeletonHorseTrapEvent;
import com.destroystokyo.paper.event.player.PlayerPickupExperienceEvent;
import net.kyori.adventure.text.Component;
import net.onelitefeather.stardust.StardustPlugin;
import net.onelitefeather.stardust.user.User;
import net.onelitefeather.stardust.user.UserPropertyType;
import org.bukkit.Bukkit;
import org.bukkit.block.data.Powerable;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Item;
import org.bukkit.entity.Player;
import org.bukkit.entity.Projectile;
import org.bukkit.event.Event;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.block.Action;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.event.block.BlockPlaceEvent;
import org.bukkit.event.block.BlockReceiveGameEvent;
import org.bukkit.event.entity.*;
import org.bukkit.event.entity.EntityDamageEvent.DamageCause;
import org.bukkit.event.player.PlayerDropItemEvent;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.event.player.PlayerItemConsumeEvent;
import org.bukkit.event.raid.RaidTriggerEvent;

import java.util.Arrays;
import java.util.List;

public class PlayerVanishListener implements Listener {

    private final StardustPlugin stardustPlugin;
    private final List<DamageCause> possibleEntityAttackCauses = Arrays.asList(
            DamageCause.ENTITY_ATTACK,
            DamageCause.ENTITY_SWEEP_ATTACK,
            DamageCause.PROJECTILE
    );

    public PlayerVanishListener(StardustPlugin stardustPlugin) {
        this.stardustPlugin = stardustPlugin;
    }

    private boolean isVanished(Player player) {
        return stardustPlugin.getUserService().getVanishService().isVanished(player);
    }

    @EventHandler
    public void handleBlockBreak(BlockBreakEvent event) {
        Player player = event.getPlayer();
        User user = stardustPlugin.getUserService().getUser(player.getUniqueId());
        if (user == null) return;
        event.setCancelled(isVanished(player) && !user.isInBuildMode());
    }

    @EventHandler
    public void handleBlockPlace(BlockPlaceEvent event) {
        Player player = event.getPlayer();
        User user = stardustPlugin.getUserService().getUser(player.getUniqueId());
        if (user == null) return;
        event.setCancelled(isVanished(player) && !user.isInBuildMode());
    }

    @EventHandler
    public void handleEntityDamage(EntityDamageEvent event) {
        if (possibleEntityAttackCauses.contains(event.getCause())) return;
        if (event.getEntity() instanceof Player player) {
            User user = stardustPlugin.getUserService().getUser(player.getUniqueId());
            boolean isDamageAllowed = user != null && user.isPvPAllowed();
            event.setCancelled(isVanished(player) && !isDamageAllowed);
        }
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    public void onEntityDamageByEntity(EntityDamageByEntityEvent event) {
        Entity target = event.getEntity();
        Entity attacker = event.getDamager();
        if (attacker instanceof Projectile) return;

        User attackerUser = stardustPlugin.getUserService().getUser(attacker.getUniqueId());
        if (attackerUser == null) return;
        User targetUser = stardustPlugin.getUserService().getUser(target.getUniqueId());

        boolean canAttack = attackerUser.isVanished() && attackerUser.isPvPAllowed();

        boolean cancel;
        if (attacker instanceof Player player && isVanished(player)) {
            cancel = !canAttack;
        } else if (targetUser != null && targetUser.isVanished()) {
            cancel = !canAttack;
        } else if (!(target instanceof Player)) {
            cancel = canAttack;
        } else if (target.isInvulnerable()) {
            cancel = !attacker.hasPermission("stardust.bypass.damage.invulnerable");
        } else {
            cancel = false;
        }
        event.setCancelled(cancel);
    }

    @EventHandler
    public void handleProjectileLaunch(ProjectileLaunchEvent event) {

        Projectile projectile = event.getEntity();
        if (!(projectile.getShooter() instanceof Player player)) return;

        User user = stardustPlugin.getUserService().getUser(player.getUniqueId());
        if (user == null) return;

        event.setCancelled(user.isVanished() && !user.isPvPAllowed());
    }

    @EventHandler
    public void handleProjectileEntityHit(ProjectileHitEvent event) {

        if (event.getHitEntity() == null) return;
        Projectile projectile = event.getEntity();

        if (!(projectile.getShooter() instanceof Player shooter)) return;

        User user = stardustPlugin.getUserService().getUser(shooter.getUniqueId());
        if (user == null) return;

        event.setCancelled(user.isVanished() && !user.isPvPAllowed());
    }

    @EventHandler
    public void onTarget(EntityTargetEvent event) {
        User user = null;
        Entity entity = null;
        if (event.getEntity() instanceof Player) {
            user = stardustPlugin.getUserService().getUser(event.getEntity().getUniqueId());
            entity = event.getEntity();
        } else if (event.getTarget() instanceof Player) {
            user = stardustPlugin.getUserService().getUser(event.getTarget().getUniqueId());
            entity = event.getTarget();
        }
        if ((user != null && user.isVanished()) || (entity != null && entity.isInvulnerable())) {
            event.setCancelled(true);
        }
    }

    @EventHandler
    public void onItemConsume(PlayerItemConsumeEvent event) {
        var player = event.getPlayer();
        User user = stardustPlugin.getUserService().getUser(player.getUniqueId());
        if (user == null) return;
        if (!user.isVanished()) return;
        event.setCancelled(true);
    }

    @EventHandler
    public void onFoodLevelChange(FoodLevelChangeEvent event) {
        if (!(event.getEntity() instanceof Player player)) return;
        if (event.getFoodLevel() >= event.getEntity().getFoodLevel()) return; //Ignore food level increment
        if (!player.isInvulnerable()) return;
        event.setCancelled(true);
    }

    @EventHandler
    public void onPickUp(EntityPickupItemEvent event) {
        if (event.getEntity() instanceof Player player) {
            try {
                User user = stardustPlugin.getUserService().getUser(player.getUniqueId());
                if (user == null) return;
                event.setCancelled(user.isVanished() && !user.isItemCollectDisabled());
            } catch (Exception e) {
                stardustPlugin.getLogger().throwing(PlayerVanishListener.class.getSimpleName(), "onPickUp", e);
            }
        }
    }

    @EventHandler
    public void onDrop(PlayerDropItemEvent event) {
        Player player = event.getPlayer();
        try {
            User user = stardustPlugin.getUserService().getUser(player.getUniqueId());
            if (user == null) return;
            event.setCancelled(user.isVanished() && !user.isItemDropDisabled());
        } catch (Exception e) {
            stardustPlugin.getLogger().throwing(PlayerVanishListener.class.getSimpleName(), "onDrop", e);
        }
    }

    @EventHandler
    public void onPlayerPickupExp(PlayerPickupExperienceEvent event) {
        Player player = event.getPlayer();
        User user = stardustPlugin.getUserService().getUser(player.getUniqueId());
        if (user == null) return;
        try {
            event.setCancelled(user.isVanished() && !user.isItemCollectDisabled());
        } catch (Exception e) {
            stardustPlugin.getLogger().throwing(PlayerVanishListener.class.getSimpleName(), "onPlayerPickupExp", e);
        }
    }

    @EventHandler
    public void onPlayerDeath(PlayerDeathEvent event) {
        Player player = event.getEntity();
        try {
            User user = stardustPlugin.getUserService().getUser(player.getUniqueId());
            if (user == null) return;
            if (user.isVanished()) {
                event.getDrops().clear();
                event.deathMessage(Component.empty());
                event.setKeepInventory(true);
                event.setKeepLevel(true);
                event.setShouldDropExperience(false);
                event.setShouldPlayDeathSound(false);
            }
        } catch (Exception e) {
            stardustPlugin.getLogger().throwing(PlayerVanishListener.class.getSimpleName(), "onPlayerDeath", e);
        }
    }

    @EventHandler
    public void handlePhysicalInteract(PlayerInteractEvent event) {
        Player player = event.getPlayer();
        if (event.getClickedBlock() == null) return;
        User user = stardustPlugin.getUserService().getUser(player.getUniqueId());
        if (user == null || !user.isVanished()) return;

        boolean isBuildingDenied = Boolean.FALSE.equals(
                user.getProperty(UserPropertyType.VANISH_ALLOW_BUILDING).getValue()
        );
        boolean isBlockPowerable = event.getClickedBlock().getBlockData() instanceof Powerable;

        if (event.getAction() == Action.PHYSICAL) {
            boolean isBlockPhysical = stardustPlugin.getPluginConfiguration().isPhysicalBlock(event.getClickedBlock().getType());
            event.setCancelled((isBlockPhysical || isBlockPowerable) ? isBuildingDenied : event.useInteractedBlock() == Event.Result.DENY);
            return;
        }
        event.setCancelled(isBlockPowerable ? isBuildingDenied : event.useInteractedBlock() == Event.Result.DENY);
    }

    @EventHandler
    public void handleGameEvent(BlockReceiveGameEvent event) {
        Player resultPlayer = null;
        if (event.getEntity() instanceof Player player) {
            resultPlayer = player;
        } else if (event.getEntity() instanceof Item item && item.getThrower() != null) {
            resultPlayer = Bukkit.getPlayer(item.getThrower());
        }

        if (resultPlayer == null) return;
        event.setCancelled(stardustPlugin.getUserService().getVanishService().isVanished(resultPlayer));
    }

    @EventHandler
    public void handleRaidTrigger(RaidTriggerEvent event) {
        Player player = event.getPlayer();
        event.setCancelled(isVanished(player));
    }

    @EventHandler
    public void handleRaidTrigger(SkeletonHorseTrapEvent event) {
        long notVanishedSize = event.getEligibleHumans().stream()
                .filter(Player.class::isInstance)
                .map(Player.class::cast)
                .filter(p -> !isVanished(p))
                .count();
        event.setCancelled(notVanishedSize == 0);
    }
}

File: src/main/kotlin/net/onelitefeather/stardust/listener/PlayerVanishListener.kt
package net.onelitefeather.stardust.listener

import com.destroystokyo.paper.event.entity.SkeletonHorseTrapEvent
import com.destroystokyo.paper.event.player.PlayerPickupExperienceEvent
import net.kyori.adventure.text.Component
import net.onelitefeather.stardust.StardustPlugin
import net.onelitefeather.stardust.user.User
import net.onelitefeather.stardust.user.UserPropertyType
import net.onelitefeather.stardust.util.PlayerUtils
import org.bukkit.Bukkit
import org.bukkit.block.data.Powerable
import org.bukkit.entity.*
import org.bukkit.event.Event
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.block.Action
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.event.block.BlockPlaceEvent
import org.bukkit.event.block.BlockReceiveGameEvent
import org.bukkit.event.entity.*
import org.bukkit.event.entity.EntityDamageEvent.DamageCause
import org.bukkit.event.player.PlayerDropItemEvent
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.event.raid.RaidTriggerEvent


class PlayerVanishListener(private val stardustPlugin: StardustPlugin) : Listener, PlayerUtils {

    private val vanishService = stardustPlugin.userService.playerVanishService
    private val possibleEntityAttackCauses = listOf(
        DamageCause.ENTITY_ATTACK,
        DamageCause.ENTITY_SWEEP_ATTACK,
        DamageCause.PROJECTILE
    )

    @EventHandler
    fun handleBlockBreak(event: BlockBreakEvent) {
        val player = event.player
        val user = stardustPlugin.userService.getUser(player.uniqueId) ?: return
        event.isCancelled = vanishService.isVanished(player) && !user.isBuildingAllowed()
    }

    @EventHandler
    fun handleBlockPlace(event: BlockPlaceEvent) {
        val player = event.player
        val user = stardustPlugin.userService.getUser(player.uniqueId) ?: return
        event.isCancelled = vanishService.isVanished(player) && !user.isBuildingAllowed()
    }

    @EventHandler
    fun handleEntityDamage(event: EntityDamageEvent) {
        if (possibleEntityAttackCauses.contains(event.cause)) return
        if (event.entity is Player) {

            val player = event.entity as Player
            val user = stardustPlugin.userService.getUser(player.uniqueId)
            val isDamageAllowed = user?.isPvPAllowed() ?: false

            event.isCancelled = vanishService.isVanished(player) && !isDamageAllowed
        }
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    fun onEntityDamageByEntity(event: EntityDamageByEntityEvent) {

        val target = event.entity
        val attacker = event.damager
        if (attacker is Projectile) return

        val attackerUser = stardustPlugin.userService.getUser(attacker.uniqueId) ?: return
        val targetUser = stardustPlugin.userService.getUser(target.uniqueId)

        val canAttack = attackerUser.isVanished() && attackerUser.isPvPAllowed()

        event.isCancelled = when {
            attacker is Player && vanishService.isVanished(attacker) -> !canAttack
            targetUser != null && targetUser.isVanished() -> !canAttack
            target !is Player -> canAttack
            target.isInvulnerable -> !attacker.hasPermission("stardust.bypass.damage.invulnerable")
            else -> false
        }
    }

    @EventHandler
    fun handleProjectileLaunch(event: ProjectileLaunchEvent) {
        if (event.entity !is ThrownPotion) return
        val entity = event.entity
        if (entity.shooter !is Entity) return
        val shooter = entity.shooter as Entity
        val user = stardustPlugin.userService.getUser(shooter.uniqueId) ?: return
        event.isCancelled = user.isVanished() && !user.isPvPAllowed()
    }

    @EventHandler
    fun handleProjectileEntityHit(event: ProjectileHitEvent) {
        if (event.hitEntity == null) return
        val entity = event.entity
        if (entity.shooter !is Entity) return
        val shooter = entity.shooter as Entity
        val user = stardustPlugin.userService.getUser(shooter.uniqueId) ?: return
        event.isCancelled = user.isVanished() && !user.isPvPAllowed()
    }

    @EventHandler
    fun onTarget(event: EntityTargetEvent) {

        var user: User? = null
        var entity: Entity? = null

        if (event.entity is Player) {
            user = stardustPlugin.userService.getUser(event.entity.uniqueId)
            entity = event.entity
        } else if (event.target is Player) {
            user = stardustPlugin.userService.getUser((event.target as Player).uniqueId)
            entity = event.target
        }

        if (user != null && user.isVanished() || entity != null && entity.isInvulnerable) {
            event.isCancelled = true
        }
    }

    @EventHandler
    fun onFoodLevelChange(event: FoodLevelChangeEvent) {
        if (event.entity is Player) {
            val player = event.entity as Player
            try {
                val user = stardustPlugin.userService.getUser(player.uniqueId)
                event.isCancelled = user != null && (user.isVanished() || player.isInvulnerable)
            } catch (e: Exception) {
                this.stardustPlugin.logger
                    .throwing(PlayerVanishListener::class.java.simpleName, "onFoodLevelChange", e)
            }
        }
    }

    @EventHandler
    fun onPickUp(event: EntityPickupItemEvent) {
        if (event.entity is Player) {
            val player = event.entity as Player
            try {
                val user = stardustPlugin.userService.getUser(player.uniqueId) ?: return
                event.isCancelled = user.isVanished() && !user.isItemCollectDisabled()
            } catch (e: Exception) {
                this.stardustPlugin.logger
                    .throwing(PlayerVanishListener::class.java.simpleName, "onPickUp", e)
            }
        }
    }

    @EventHandler
    fun onDrop(event: PlayerDropItemEvent) {
        val player = event.player
        try {
            val user = stardustPlugin.userService.getUser(player.uniqueId) ?: return
            event.isCancelled = user.isVanished() && !user.isItemDropDisabled()
        } catch (e: Exception) {
            this.stardustPlugin.logger
                .throwing(PlayerVanishListener::class.java.simpleName, "onDrop", e)
        }
    }

    @EventHandler
    fun onPlayerPickupExp(event: PlayerPickupExperienceEvent) {
        val player = event.player
        val user = stardustPlugin.userService.getUser(player.uniqueId) ?: return
        try {
            event.isCancelled = user.isVanished() && !user.isItemCollectDisabled()
        } catch (e: Exception) {
            this.stardustPlugin.logger
                .throwing(PlayerVanishListener::class.java.simpleName, "onPlayerPickupExp", e)
        }
    }

    @EventHandler
    fun onPlayerDeath(event: PlayerDeathEvent) {
        val player = event.entity
        try {
            val user = stardustPlugin.userService.getUser(player.uniqueId) ?: return
            if (user.isVanished()) {
                event.drops.clear()
                event.deathMessage(Component.empty())
                event.keepInventory = true
                event.keepLevel = true
                event.setShouldDropExperience(false)
                event.setShouldPlayDeathSound(false)
            }
        } catch (e: Exception) {
            this.stardustPlugin.logger
                .throwing(PlayerVanishListener::class.java.simpleName, "onPlayerDeath", e)
        }
    }

    @EventHandler
    fun handlePhysicalInteract(event: PlayerInteractEvent) {

        val player = event.player

        val block = event.clickedBlock ?: return
        val user = stardustPlugin.userService.getUser(player.uniqueId) ?: return
        if (!user.isVanished()) return
        if (user.getProperty(UserPropertyType.VANISH_ALLOW_BUILDING)?.getValue<Boolean>() == true) return

        val isBlockPowerable = block.blockData is Powerable

        if (event.action == Action.PHYSICAL) {
            val isBlockPhysical = stardustPlugin.pluginConfig.physicalBlocks().contains(block.type)

            event.isCancelled = if (isBlockPhysical || isBlockPowerable) {
                true
            } else {
                event.useInteractedBlock() == Event.Result.DENY
            }

            return
        }

        event.isCancelled = if (isBlockPowerable) {
            !player.isSneaking
        } else {
            event.useInteractedBlock() == Event.Result.DENY
        }
    }

    @EventHandler
    fun handleGameEvent(event: BlockReceiveGameEvent) {

        val player = if (event.entity is Player) {
            event.entity as Player
        } else if (event.entity is Item) {
            val thrower = (event.entity as Item).thrower
            if (thrower != null) Bukkit.getPlayer(thrower) else null
        } else {
            null
        }

        if (player == null) return
        event.isCancelled = stardustPlugin.userService.playerVanishService.isVanished(player)
    }

    @EventHandler
    fun handleRaidTrigger(event: RaidTriggerEvent) {
        val player = event.player
        event.isCancelled = vanishService.isVanished(player)
    }

    @EventHandler
    fun handleRaidTrigger(event: SkeletonHorseTrapEvent) {
        val notVanishedSize = event.eligibleHumans.filterIsInstance<Player>().filterNot(vanishService::isVanished).size
        event.isCancelled = notVanishedSize == 0
    }
}